### PR TITLE
feat: complete `UndelegateAndTransfer` and `DepositAndDelegate` flows

### LINKED ISSUE #10046: staking-combinations.contract.js is incomplete
## What is the Problem Being Solved?

`staking-combinations.contract.js` is the only contract demonstrating using flows for continuing offers, but is incomplete.

## Description of the Design

Complete the `depositAndDelegate` and `undelegateAndTransfer` flows.

### PR DESCRIPTION
**Pull request by 0xpatrickdev**
closes: #10046

## Description
- use `AnyNatAmountShape` for `AmountArg` in `@agoric/orchestration`
- Update  `undelegate` to take `AmountArg` and `ChainAddress`
- Update `undelegate` to accept and `bondDenom`
- Finish `UnelegateAndTransfer` and `DepositAndDelegate` flows in staking-combinations contract

### Security Considerations
n/a

### Scaling Considerations
n/a

### Documentation Considerations
n/a

### Testing Considerations
Updates existing tests and includes tests for new functionality

### Upgrade Considerations
n/a

### CODE REVIEW

File: packages/async-flow/src/types.d.ts [hunk 1]
```diff
@@ -68,9 +68,11 @@ export type GuestInterface<T> = {
     ? (...args: Parameters<T[K]>) => Promise<R>
     : T[K] extends HostAsyncFuncWrapper
       ? GuestOf<T[K]>
-      : T[K] extends object
-        ? GuestInterface<T[K]>
-        : T[K];
+      : T[K] extends (...args: any[]) => infer R
+        ? (...args: Parameters<T[K]>) => R
```
Comment (turadg) [inline packages/async-flow/src/types.d.ts]: Isn' this just:
```suggestion
      : T[K] extends (...args: any[]) => infer R
        ? T[K]
```
I don't think that's quite right though. What failure did you run into? Please add the cases to `async-flow/test/types.test-d.ts` so we can avoid regressions in these types.

File: packages/orchestration/USAGE.md [hunk 2]
```diff
@@ -17,7 +17,7 @@ See [`src/examples`](src/examples)
...
 | [stakeBld](/packages/orchestration/src/examples/stakeBld.contract.js) | Ready 🟢 | Returns a `LocalOrchestrationAccount` that can perform staking actions. | - `LocalOrchestrationAccount` | Ready 🟢 |
 | [stakeIca](/packages/orchestration/src/examples/stakeIca.contract.js) | Ready 🟢 | Returns a `CosmosOrchestrationAccount` that can perform staking actions. | - `CosmosOrchestrationAccount` | Ready 🟢 |
-| [staking-combinations](/packages/orchestration/src/examples/staking-combinations.contract.js) | Under Construction 🚧 | Combines actions into a single offer flow and demonstrates writing continuing offers. | - `CosmosOrchestrationAccount`<br>- `CombineInvitationMakers` <br>- Continuing Offers |
+| [staking-combinations](/packages/orchestration/src/examples/staking-combinations.contract.js) | Ready 🟢 | Combines actions into a single offer flow and demonstrates writing continuing offers. | - `CosmosOrchestrationAccount`<br>- `CombineInvitationMakers` <br>- Continuing Offers |
```
Comment (turadg) [inline packages/orchestration/USAGE.md:20]: 🎉

File: packages/orchestration/src/examples/staking-combinations.contract.js [hunk 3]
```diff
@@ -67,27 +86,39 @@ const contract = async (zcf, privateArgs, zone, { orchestrateAll }) => {
...
-            orchFns.depositAndDelegate(account, seat, validatorAddr, amountArg),
+            orchFns.depositAndDelegate(account, seat, validator),
           'Deposit and delegate',
           undefined,
           {
             give: {
               Stake: AmountShape,
             },
+            want: {},
+            // user cannot exit their seat; contract must exit it.
```
Comment (turadg) [inline packages/orchestration/src/examples/staking-combinations.contract.js:97]: ty

File: packages/orchestration/src/examples/staking-combinations.contract.js [hunk 4]
```diff
@@ -104,12 +135,16 @@ const contract = async (zcf, privateArgs, zone, { orchestrateAll }) => {
...
   const orchFns = orchestrateAll(flows, {
+    contractState,
     makeCombineInvitationMakers,
     makeExtraInvitationMaker,
     flows,
     zcf,
+    zoeTools,
   });
 
+  const creatorFacet = prepareChainHubAdmin(zone, chainHub);
```
Comment (turadg) [inline packages/orchestration/src/examples/staking-combinations.contract.js:144]: since this is an example contract, please include a comment about why this is needed.

File: packages/orchestration/src/examples/staking-combinations.flows.js [hunk 5]
```diff
@@ -41,43 +45,65 @@ harden(makeAccount);
...
   account,
   seat,
   validator,
-  amount,
 ) => {
-  console.log('depositAndDelegate', account, seat, validator, amount);
-  // TODO deposit the amount
-  await account.delegate(validator, amount);
+  await null;
+  console.log('depositAndDelegate', account, seat, validator);
```
Comment (turadg) [inline packages/orchestration/src/examples/staking-combinations.flows.js]: please use a tracer for the module so the log messages have a namespace

File: packages/orchestration/src/examples/staking-combinations.flows.js [hunk 6]
```diff
@@ -41,43 +45,65 @@ harden(makeAccount);
...
+ *   delegations: { amount: AmountArg; validator: CosmosValidatorAddress }[];
+ *   destination: ChainAddress;
+ * }} offerArgs
  * @returns {Promise<string>}
  */
 export const undelegateAndTransfer = async (
-  orch,
-  ctx,
+  _orch,
+  _ctx,
```
Comment (turadg) [inline packages/orchestration/src/examples/staking-combinations.flows.js]: what prompted this renaming? let's only use `_` when linting requires it. It's obvious enough in the editor that it's never used. (same applies to `_privateArgs` change in the contract)

File: packages/orchestration/src/exos/chain-hub-admin.js [hunk 7]
```diff
@@ -37,7 +37,7 @@ export const prepareChainHubAdmin = (zone, chainHub) => {
        * @param {CosmosChainInfo} chainInfo
        * @param {IBCConnectionInfo} connectionInfo
        */
-      async initChain(chainName, chainInfo, connectionInfo) {
+      async registerChain(chainName, chainInfo, connectionInfo) {
```
Comment (turadg) [inline packages/orchestration/src/exos/chain-hub-admin.js:53]: better name. I forget why we had `init` before. Maybe thinking it would be only allowed once? but clearly what it's doing is registering

File: packages/orchestration/src/typeGuards.js [hunk 8]
```diff
@@ -37,11 +37,17 @@ export const Proto3Shape = {
   value: M.string(),
 };
 
-// FIXME missing `delegatorAddress` from the type
+export const DenomShape = M.string();
+
+/** @type {TypedPattern<DenomAmount>} */
+export const DenomAmountShape = { denom: DenomShape, value: M.bigint() };
+
+export const AmountArgShape = M.or(NatAmountShape, DenomAmountShape);
```
Comment (turadg) [inline packages/orchestration/src/typeGuards.js]: I think we're meant to support non-nat amounts, but the `orchestration-api.ts` is this narrow so this is consistent with that. FYI @dtribble 

Still I would suggest leaving this as `AmountShape` for future compatibility and so that we don't need a NatAmountShape` export from `@agoric/ertp`. It's better to use `makeNatAmountShape`, which we already have an export for, because it validates that particular brand. If people see the generic one they'll reach for that and in most cases they're better off with the narrower one.

File: packages/orchestration/src/examples/staking-combinations.flows.js (hunk 6 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/examples/staking-combinations.flows.js]: Why isn't the linter upset, the one in my brain is. I thought we [error](https://github.com/Agoric/agoric-sdk/blob/c9fb1e70f2846b75e113a34a52525cb96a203337/packages/eslint-config/eslint-config.cjs#L34-L40) on `no-used-vars` with `argsIgnorePattern: '^_',`?

File: packages/orchestration/src/typeGuards.js (hunk 8 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/typeGuards.js]: Wouldn't we have a new guard when it comes time to support non-nat amounts? Looking at [ics-721-nft-transfer//README.md](https://github.com/cosmos/ibc/blob/da6b0728bffb348daea3964dedf124a44a5fed27/spec/app/ics-721-nft-transfer/README.md), I don't see `denom` mentioned. It uses different identifiers for the token (`classId` and `tokenId`).

I am now seeing `makeNatAmountShape` and agree that' what we should document for contract authors. For the OrchAccount exos, it seems we should still narrow to `NatAmount` as none of what's been built will work with other ERTP asset types. Using `makeNatAmountShape` for this seems challenging as there are many acceptable brands.

File: packages/async-flow/src/types.d.ts (hunk 1 above)
Comment (0xpatrickdev) [reply packages/async-flow/src/types.d.ts]: Thanks, this is simpler and works. The error I was seeing was: `Type 'GuestInterface<() => HostInterface<ChainAddress>>' has no call signatures.`

I added a type test in [f10bad5](https://github.com/Agoric/agoric-sdk/pull/10045/commits/f10bad5eba688629789f94ad19b9679d91654c92). 

I was expecting to be able to do:
```diff
@@ -1,5 +1,6 @@
 import { expectType } from 'tsd';
 import type { Vow, VowTools } from '@agoric/vow';
+import type { Guarded } from '@endo/exo';
 import type {
   HostOf,
   GuestOf,
@@ -30,6 +31,7 @@ type ExoAPIBase = {
   getValue: () => number;
   setValue: (value: number) => void;
   getCopyData: () => Record<string, number>[];
+  getRemote: () => Guarded<{ doFoo: () => void }>;
 };
 type ExoGuestAPI = ExoAPIBase & {
   getValueAsync: () => Promise<number>;
```

But see:
```
Type 'HostInterface<Guarded<{ doFoo: () => void; }>>' is missing the following properties from type 'RemotableBrand<{}, { doFoo: () => void; } & GetInterfaceGuard<{ doFoo: () => void; }>>': L, R
```

Is `Guarded` the correct way to type a durable remotable?

File: packages/orchestration/src/examples/staking-combinations.flows.js (hunk 6 above)
Comment (turadg) [reply packages/orchestration/src/examples/staking-combinations.flows.js]: we do, to not have extra bindings hanging around. It's usually indicative of a coding error.

but when it's in a list of params and you do use the latter params, binding the earlier ones is necessary. so there's no point in warning someone that they didn't use them

File: packages/orchestration/src/typeGuards.js (hunk 8 above)
Comment (turadg) [reply packages/orchestration/src/typeGuards.js]: > For the OrchAccount exos, it seems we should still narrow to NatAmount 

Agree but I don't think that justifies exporting it from `@agoric/ertp`. You could just define it here:
```js
/** @see {makeNatAmountShape} to match a specific brand */
const AnyNatAmountShape = {brand: BrandShape, value: NatShape};
export const AmountArgShape = M.or(AnyNatAmountShape, DenomAmountShape);
```

File: packages/orchestration/src/examples/staking-combinations.flows.js (hunk 6 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/examples/staking-combinations.flows.js]: Ah I see now - didn't realize this was the case but makes sense. Thanks for explaining!

File: packages/async-flow/test/types.test-d.ts [hunk 9]
```diff
@@ -20,3 +24,48 @@ expectType<(p1: number, p2: number) => Vow<void>>(
...
+expectType<
+  ExoAPIBase & {
+    getValueAsync: () => Vow<number>;
+  }
+>(null as unknown as HostInterface<ExoGuestAPI>);
+expectType<
+  ExoAPIBase & {
+    getValueAsync: () => Promise<number>;
+  }
+>(null as unknown as GuestInterface<ExoHostAPI>);
```
Comment (turadg) [inline packages/async-flow/test/types.test-d.ts]: I see `null as unknown as Type` was here before but it's getting a little hard to parse. consider,
```ts
const castable: unknown = null;
```

**Review by turadg — APPROVED** [review]

File: packages/async-flow/test/types.test-d.ts (hunk 9 above)
Comment (0xpatrickdev) [reply packages/async-flow/test/types.test-d.ts]: Cool will do. Any thoughts on how to mock a remotable? https://github.com/Agoric/agoric-sdk/pull/10045#discussion_r1751074827

File: packages/async-flow/src/types.d.ts (hunk 1 above)
Comment (turadg) [reply packages/async-flow/src/types.d.ts]: Hmm, `Guarded` should:
```
export type Farable<M extends Methods> = M &
  RemotableBrand<{}, M> &
  RemotableObject;
export type Guarded<M extends Methods> = Farable<M & GetInterfaceGuard<M>>;
```
but RemotableBrand expects these private properties,
```
export declare class RemotableBrand<Local, Remote> {
  /** The local properties of the object. */
  private L: Local;

  /** The type of all the remotely-callable functions. */
  private R: Remote;
}
```

I suppose we need to fix `RemotableBrand` so it doesn't appear to have `L` and `R` as real properties. WDYT @michaelfig ? 

Meanwhile I think it's okay to cast or expect error with an explanation

Comment (michaelfig) [reply packages/async-flow/src/types.d.ts]: Please don't change `RemotableBrand`... it's carefully tuned to give proper errors when you pass it an object that wasn't created correctly as a Remotable.  Consider it an internal type that makes `E` work with things like Exo-objects, Remotable, Far, and durables, not for use outside that layer.

To me, it seems that `HostInterface` is at fault; it needs to explicitly handle any RemotableBrand that was implied by its template parameter.

Comment (michaelfig) [reply packages/async-flow/src/types.d.ts]: > Isn't this just:

Not the same, but I'm not sure what was intended.

The more complex version returned the type of just the function itself (such as the type of a simple wrapper function), but the "simpler" version preserved the type of the function including all the subproperties that might be hanging off of it.

Comment (0xpatrickdev) [reply packages/async-flow/src/types.d.ts]: > Meanwhile I think it's okay to cast or expect error with an explanation

I just left a TODO for now

> but I'm not sure what was intended

The changes were to support: `getChainAddress: () => ChainAddress`
